lib/scala/downloader/src/main/scala/org/enso/downloader/http/URIBuilder.scala

```diff
@@ -1,33 +1,42 @@
 package org.enso.downloader.http
 
-import akka.http.scaladsl.model.Uri
+import java.net.URI
 
 /** A simple immutable builder for URIs based on URLs.
   *
   * It contains very limited functionality that is needed by the APIs used in
   * the launcher. It can be easily extended if necessary.
   */
-case class URIBuilder private (uri: Uri) {
+case class URIBuilder private (uri: URI) {
 
   /** Resolve a segment over the path in the URI.
     *
     * For example adding `bar` to `http://example.com/foo` will result in
     * `http://example.com/foo/bar`.
     */
   def addPathSegment(segment: String): URIBuilder = {
-    val part = "/" + segment
-    copy(uri.withPath(uri.path + part))
+    val pathItems = uri.getRawPath.split("/")
+    val newPath   = (pathItems :+ segment).mkString("/")
+    copy(uri.resolve(newPath))
   }
 
   /** Add a query parameter to the URI.
     *
     * The query is appended at the end.
     */
-  def addQuery(key: String, value: String): URIBuilder =
-    copy(uri.withQuery(uri.query().+:((key, value))))
+  def addQuery(key: String, value: String): URIBuilder = {
+    val scheme    = uri.getScheme
+    val authority = uri.getAuthority
+    val path      = uri.getPath
+    val query     = if (uri.getQuery == null) "" else uri.getQuery + "&"
+    val fragment  = uri.getFragment
+    val newQuery  = query + key + "=" + value
+    val newUri    = new URI(scheme, authority, path, newQuery, fragment)
+    copy(newUri)
+  }
```

Comment:
What about encoding the `key` and `value`?

Again, replacing an existing solution with a self-built one puts it on us to ensure that it works also in irregular cases.

Well, I don't want to be complaining too much - with where this downloader is currently being used I guess maybe a non-encoding query builder will work OK. But the moment a non-expected character shows up e.g. in a version number or username, it may break. That's why I think using existing solutions is better.

----

Of course not suggesting to stay with Akka, it is an overkill for sure.

But maybe we should use Apache HttpCore for these builders?
Or at least let's either do proper encoding or admit we can't handle non-standard characters and add a check that the key/value only contains `[A-Za-z0-9-_]` or something like that. If stuff breaks, at least it will be clearer why.

Reply:
Btw. we do have our own classes for handling these.

I actually did break what I say here that 'I prefer to use existing solutions instead of maintaining our own' because there was pressure to try avoiding the Apache Http as dependency of `Standard.Base`.

But we do have tests for lots of various edge cases there.

---

Maybe you could refactor the common parts (perhaps `URITransformer`) into a separate library that can be used from both places?